src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -467,6 +469,76 @@ void four_index_constraint(
     }
   }
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+Scalar<DataType> constraint_energy(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& f_constraint,
+    const tnsr::ia<DataType, SpatialDim, Frame>& two_index_constraint,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& three_index_constraint,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& four_index_constraint,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    double gauge_constraint_multiplier, double two_index_constraint_multiplier,
+    double three_index_constraint_multiplier,
+    double four_index_constraint_multiplier) noexcept {
+  auto energy = make_with_value<Scalar<DataType>>(gauge_constraint, 0.0);
+  constraint_energy<SpatialDim, Frame, DataType>(
+      &energy, gauge_constraint, f_constraint, two_index_constraint,
+      three_index_constraint, four_index_constraint, inverse_spatial_metric,
+      gauge_constraint_multiplier, two_index_constraint_multiplier,
+      three_index_constraint_multiplier, four_index_constraint_multiplier);
+  return energy;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void constraint_energy(
+    gsl::not_null<Scalar<DataType>*> energy,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& f_constraint,
+    const tnsr::ia<DataType, SpatialDim, Frame>& two_index_constraint,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& three_index_constraint,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& four_index_constraint,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    double gauge_constraint_multiplier, double two_index_constraint_multiplier,
+    double three_index_constraint_multiplier,
+    double four_index_constraint_multiplier) noexcept {
+  if (get_size(get(*energy)) != get_size(get<0>(f_constraint))) {
+    *energy = Scalar<DataType>{get<0>(f_constraint)};
+  }
```

Comment:
I don't quite understand what you are doing here. This is similar to the code from #1172 where we commented
https://github.com/sxs-collaboration/spectre/pull/1172#discussion_r245938351

Is this basically allocating `energy`?

Reply:
There's no guarantee that energy points to something of the correct size. This checks that this size is correct and allocates something of the correct size if not.

Reply:
This is checking that *energy has the correct size and if not, allocating something that does have the correct size.